lms/resources/canvas_oauth2_redirect.py

```diff
@@ -0,0 +1,24 @@
+from pyramid.security import Allow
+
+from lms.resources._js_config import JSConfig
+
+
+class CanvasOAuth2RedirectResource:
+    """Resource for the Canvas OAuth 2 redirect popup."""
+
+    __acl__ = [(Allow, "lti_user", "canvas_api")]
+
+    def __init__(self, request):
+        self._js_config = None
+        self._request = request
+
+    @property
+    def js_config(self):
+        if not self._request.exception:
+            # The normal views do not need to configure the frontend, only
+            # the exception views.
+            return None
+
+        if not self._js_config:
+            self._js_config = JSConfig(self, self._request)
+        return self._js_config
```

Comment:
```suggestion
    def __init__(self, request):
        self.js_config = JSConfig(self, request)
```

I think this works although some tests will need to be updated. Let's keep the code simple if we can

Reply:
The problem with the suggested approach is that it will cause the `base.html` template to load the client JS during the non-exception view, which we don't want. See https://github.com/hypothesis/lms/blob/3747a66d6691e20ef93302424285ccb38d8957ef/lms/templates/base.html.jinja2#L39

Reply:
In the context of the `oauth2_redirect_error` view this happens to work from the user's point of view because the `oauth2_redirect.html` template calls `window.close()` as soon as the page loads.

The client frontend scripts load synchronously though so the `window.close()` block will not currently run until after the frontend JS has been fetched and run. Again, these are small so the user is unlikely to notice, but it still adds a very small unnecessary delay.

Reply:
Actually, nevermind. I see that `oauth2_redirect.html` overrides the default `scripts` block so the above comment is wrong.